docs - useMetabot hook (#73692)
metabase/metabase: docs/embedding/sdk/ai-chat.md

- `auto` (default): Metabot uses the `stacked` layout on mobile screens, and a `sidebar` layout on larger screens.
- `stacked`: the question visualization stacks on top of the chat interface.
- `sidebar`: the question visualization appears to the left of the chat interface, which is on a sidebar on the right.

## Building custom AI chat UIs with `useMetabot`

If `MetabotQuestion`'s built-in layouts don't fit your app, use the `useMetabot` hook to read Metabot's conversation state directly and render your own UI. The hook gives you the messages, the chart the agent most recently produced, processing and error state, and actions to submit, cancel, retry, or reset the conversation.

### AI chat with inline charts

![AI chat inline charts](../images/ai-chat-inline-chart.png)

> Each agent chart message carries a pre-wired `Chart` component. Walk the messages and render charts inline alongside the chat transcript:

Reviewer: > agent chart message

what's that?

> Each agent chart message carries a pre-wired `Chart` component. Walk the messages and render charts inline alongside the chat transcript:

Author: ```suggestion
When an agent responds, the message can contain a `Chart` component. You can walk the agent's messages and render charts inline alongside the chat transcript:
```

```typescript
{% include_file "{{ dirname }}/snippets/questions/use-metabot-inline-charts.tsx" %}
```

### AI chat with dedicated chart panel

![AI chat dedicated chart](../images/ai-chat-dedicated-chart.png)

> The `CurrentChart` component is bound to the latest chart the agent produced. Render `CurrentChart` once, and it will swap in new charts as the agent creates them. You'll want to filter chart messages out of the transcript so they don't render twice:

Reviewer: > You'll want to filter chart messages out of the transcript so they don't render twice:

implies that the following snippet does that. But where? There's no comment in the code that addresses this part

Author: Comments are on line 65. Were they not clear?

Added the keyword "filter" so people can make the connection.

```typescript
{% include_file "{{ dirname }}/snippets/questions/use-metabot-dedicated-chart.tsx" %}
```

### Notes on `useMetabot`

- **Guard against null while waiting for the SDK bundle**: `useMetabot` returns `null` until the SDK bundle has loaded and `<MetabaseProvider>` has mounted. Always guard before use. If you don't guard it, the first render will throw `Cannot read properties of null` when you reach for `metabot.messages`, `metabot.submitMessage`, etc., because the SDK ships its Metabot internals via a code-split chunk that isn't available synchronously.

> - **Bring your own Markdown renderer**: `MetabotQuestion` renders agent text messages internally, including markdown formatting, transcript scrolling, and input styling. The `useMetabot` hook hands you the raw conversation state, which means you own the rendering. In particular, agent text messages (`message.type === 'text'`) contain **markdown**: links, bold, lists, inline code. The snippets below render `message.message` as plain text for brevity, but production usage should pass the text through a markdown renderer (`react-markdown`, `markdown-to-jsx`, or your own) so links and formatting display correctly.

Reviewer: >The snippets below

there are no snippets below

Author: Ah, right. I had this section above at one point. Nice catch.

```suggestion
- **Bring your own Markdown renderer**: `MetabotQuestion` renders agent text messages internally, including markdown formatting, transcript scrolling, and input styling. The `useMetabot` hook hands you the raw conversation state, which means you own the rendering. In particular, agent text messages (`message.type === 'text'`) contain **markdown**: links, bold, lists, inline code. The snippets above render `message.message` as plain text for brevity, but production usage should pass the text through a markdown renderer (`react-markdown`, `markdown-to-jsx`, or your own) so links and formatting display correctly.
```

> - **Strip links returned by the agent**: the agent text may include links pointing back to the host Metabase (like a link to a chart it created). Those links require an authenticated Metabase session, so people won't be able to view the links.

Reviewer: aren't people authenticated into Metabase when they're using the AI chat?

Reviewer: do the code snippets do that?

Author: Code snippets show the `defineMetabaseAuthConfig` call.

- **Bring your own Markdown renderer**: `MetabotQuestion` renders agent text messages internally, including markdown formatting, transcript scrolling, and input styling. The `useMetabot` hook hands you the raw conversation state, which means you own the rendering. In particular, agent text messages (`message.type === 'text'`) contain **markdown**: links, bold, lists, inline code. The snippets below render `message.message` as plain text for brevity, but production usage should pass the text through a markdown renderer (`react-markdown`, `markdown-to-jsx`, or your own) so links and formatting display correctly.
- **Strip links returned by the agent**: the agent text may include links pointing back to the host Metabase (like a link to a chart it created). Those links require an authenticated Metabase session, so people won't be able to view the links.

> - **The chart components accept different props depending on `drills`**: both the `Chart` on each chart message and the `CurrentChart` returned by the hook are pre-wired with `questionId`, `token`, and `query`, so you shouldn't pass those props. With the default `drills={false}`, the component renders [`StaticQuestion`](./questions.md#staticquestion) and accepts its remaining props. With `drills={true}`, the component renders [`InteractiveQuestion`](./questions.md#interactivequestion) and accepts its remaining props.

Reviewer 2: I don't want to have to go that far and say they're prewired with `questionId`, `token`, and `query` because that is very technical and is also incorrect.

Correction: It's only prewired with `query`, but you can think about this 3 props as they provide some kinds of ID for the question. Of these 3 props, only one can be passed at a time. We have TypeScript to catch that, but if users pass more, our question would likely fail with weird errors.

### Inline charts

![AI chat inline charts](../images/ai-chat-inline-chart.png)

Each agent chart message carries a pre-wired `Chart` component. Walk the messages and render charts inline alongside the chat transcript:

```typescript
{% include_file "{{ dirname }}/snippets/questions/use-metabot-inline-charts.tsx" %}
```

### Dedicated chart panel

![AI chat dedicated chart](../images/ai-chat-dedicated-chart.png)

`CurrentChart` is a pre-wired component bound to the latest chart the agent produced. Render it once on the side and it swaps as new charts arrive. Filter chart messages out of the transcript so they don't render twice:

```typescript
{% include_file "{{ dirname }}/snippets/questions/use-metabot-dedicated-chart.tsx" %}
```

### Return shape

{% include_file "{{ dirname }}/api/snippets/UseMetabotResult.md" snippet="properties" %}

### Notes on `useMetabot`

> - **Guard against null while waiting for the SDK bundle**: `useMetabot` returns `null` until the SDK bundle has loaded and `<MetabaseProvider>` has mounted. Always guard before use. If you don't guard it, the first render will throw `Cannot read properties of null` when you reach for `metabot.messages`, `metabot.submitMessage`, etc., because the SDK ships its Metabot internals via a code-split chunk that isn't available synchronously.

Reviewer 2: 👍

> - **Bring your own Markdown renderer** `MetabotQuestion` renders agent text messages internally, including markdown formatting, transcript scrolling, and input styling. The `useMetabot` hook hands you the raw conversation state. Which means you own the rendering. In particular, agent text messages (`message.type === 'text'`) contain **markdown**: links, bold, lists, inline code. The snippets below render `message.message` as plain text for brevity, but production usage should pass the text through a markdown renderer (`react-markdown`, `markdown-to-jsx`, or your own) so links and formatting display correctly.

Reviewer 2: Not sure if it's worth calling out to users, but the link is for a Metabase instance e.g. `/question#base64` when rendered with markdown, it will most likely just lead to nothing. Even if users have configured the markdown renderer with a base URL (if any). SDK end users will still have to have a credential to the Metabase instance to view the question.

We have a follow-up task for this, but it's not prioritized yet.  We think ideally the link should be stripped from Embedding's Metabot https://linear.app/metabase/issue/EMB-1639/strip-link-url-from-usemetabot-responses

> - **The chart components accept the same props as `StaticQuestion`**. The `Chart` component on each chart message, and the `CurrentChart` returned by the hook, accept the same props as [`StaticQuestion`](./questions.md#staticquestion) except `questionId`, `token`, and `query`, which the SDK already wires for you. Most apps will set `drills` (to allow click-actions on chart elements) and `height`. The `height="100%"` only works if the parent has a defined height. The dedicated-chart snippet below sets `height: 600` on the outer flex container for that reason.

Reviewer 2: Just to be blunt. When `drills={true}`, we will render `InteractiveQuestion` under the hood, or `StaticQuestion` when it's false. Both components have similar props anyway, but I think we should mention this somehow.

drills: true = InteractiveQuestion props
drills: false = StaticQuestion props
